src/node/witness.rs

```diff
@@ -0,0 +1,329 @@
+
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use super::{Constructible, CoreConstructible, JetConstructible, WitnessConstructible};
+use super::{Converter, Hide, Inner, NoWitness, Node, NodeData, Redeem, RedeemData, RedeemNode};
+
+/// ID used to share [`WitnessNode`]s.
+///
+/// This is impossible to construct, which is a promise that it is impossible
+/// to share [`WitnessNode`]s.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub enum WitnessId {}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub enum Witness {}
+
+impl<J: Jet> NodeData<J> for Witness {
+    type CachedData = WitnessData<J>;
+    type Witness = Option<Arc<Value>>;
+    type SharingId = WitnessId;
+
+    fn compute_sharing_id(_: Cmr, _: &WitnessData<J>) -> Option<WitnessId> {
+        None
+    }
+}
+
+pub type WitnessNode<J> = Node<Witness, J>;
+
+impl<J: Jet> WitnessNode<J> {
+    /// Creates a copy of the node (and its entire DAG with the prune bit set)
+    pub fn pruned(&self) -> Arc<Self> {
+        let new_data = WitnessData {
+            must_prune: true,
+            ..self.data.clone()
+        };
+        Arc::new(WitnessNode {
+            data: new_data,
+            cmr: self.cmr,
+            inner: self
+                .inner
+                .as_ref()
+                .map(Arc::clone)
+                .map_witness(|wit| wit.as_ref().map(Arc::clone)),
+        })
+    }
+
+    /// Accessor for the node's arrow
+    pub fn arrow(&self) -> &Arrow {
+        &self.data.arrow
+    }
+
+    /// Whether the "must prune" bit is set on this node
+    pub fn must_prune(&self) -> bool {
+        self.data.must_prune
+    }
+
+    pub fn prune_and_retype(&self) -> Arc<Self> {
+        struct Retyper<J: Jet>(PhantomData<J>);
+
+        impl<J: Jet> Converter<Witness, Witness, J> for Retyper<J> {
+            type Error = types::Error;
+            fn convert_witness(
+                &mut self,
+                _: &PostOrderIterItem<&WitnessNode<J>>,
+                wit: &Option<Arc<Value>>,
+            ) -> Result<Option<Arc<Value>>, Self::Error> {
+                Ok(wit.as_ref().map(Arc::clone))
+            }
+
+            fn prune_case(
+                &mut self,
+                _: &PostOrderIterItem<&WitnessNode<J>>,
+                left: &Arc<WitnessNode<J>>,
+                right: &Arc<WitnessNode<J>>,
+            ) -> Result<Hide, Self::Error> {
+                // If either child is marked as pruned, we hide it, which will cause this
+                // case node to become an assertl or assertr, potentially reducing the size
+                // of types since there will be fewer constraints to unify.
+                //
+                // If both children are marked pruned, this function will only hide the left
+                // one. This doesn't matter; in this case the node itself will be marked as
+                // pruned and eventually get dropped.
```

Comment:
5b76889ffee871cb1e537cf5a2529c9400428408: I may have to run `prune_and_retype()` arbitrarily many times before I finish, right?

For every number I can construct a chain of nested `case` combinators with one `fail` at the end. In every step the lowermost `case` will be converted into `assert{l,r}` which is marked `must_prune`. The ancestor `case` combinators will not be marked because both branches must be marked `must_prune` for `case` to be marked.

This is not necessarily an issue, but we should warn users in the documentation and we should be careful about our own code.

Reply:
I think you're confused. If both branches if every `case` are prunable then they'll all be pruned in one pass. If only the topmost one is prunable, and the others have some non-prunable branches, then they won't be. Which is correct.

Reply:
You're right. Is it fair to say that an expression will be fully pruned after two steps? If you want to do anything useful with the `WitnessNode` you're going to finalize it to `RedeemNode` which does the pruning for you. So I suppose the expression is "fully pruned" after one step.

Reply:
Right -- `prune_and_retype` will do all the "explicitly-marked" pruning, in one step, and you should do this whenever you might prune stuff for optimal results. But to actually use the result, you want to do the "real" pruning, which also happens in one step, during the conversion to `RedeemNode`.